src/Substance.hs

```diff
@@ -4,535 +4,450 @@
-    subDefs :: M.Map String ([(SubType, String)], FOLExpr), -- all definitions
-    subApps :: [(FOLExpr, M.Map String String, [String])], -- Def, varmap, ids to be solved by alloy
-    subSymbols :: M.Map String SubType, -- symbol table
-    subArgs :: M.Map String [String]
-    -- subAppliedDefs :: [FOLExpr]
-} deriving (Show, Eq, Typeable)
+    return (PredicateConst i)
+
+functionParser :: Parser Func
+functionParser = do
+  n <- identifier
+  args <- parens (exprParser `sepBy1` comma)
+  return (Func { nameFunc = n, argFunc = args })
+
+exprParser, varE, applyF :: Parser Expr
+exprParser = try applyF <|> try varE
+varE = do
+  i <- varParser
+  return (VarE i)
+applyF = do
+  f <- functionParser
+  return (ApplyExpr f)
+
+predicateArgParser, predicateArgParserE, predicateArgParserP  :: Parser PredArg
+predicateArgParser = try predicateArgParserE <|> predicateArgParserP
+predicateArgParserE = do
+  e <- exprParser
+  return (PE e)
+predicateArgParserP = do
+  p <- predicateParser
+  return (PP p)
+
+predicateParser :: Parser Predicate
+predicateParser = do
+  n    <- predicateNameParser
+  args <- parens (predicateArgParser `sepBy1` comma)
+  pos  <- getPosition
+  return (Predicate { predicateName = n, predicateArgs = args, predicatePos = pos })
+
+subStmt, decl, bind, applyP :: Parser SubStmt
+subStmt = try bind <|> try decl <|> applyP
+decl = do
+  t' <- tParser
+  v' <- varParser
+  return (Decl t' v')
+bind = do
+  v' <- varParser
+  eq
+  e' <- exprParser
+  return (Bind v' e')
+applyP = do
+  p <- predicateParser
+  return (ApplyP p)
+
+----------------------------------------- Substance Typechecker ---------------------------
 
 -- | 'check' is the top-level semantic checking function. It takes a Substance
--- program as the input, checks the validity of the program, and outputs
--- a collection of information.
--- The check is done in two passes. First check all the declarations of
--- stand-alone objects like `Set`, and then check for all other things
-check :: SubProg -> SubEnv
-check p = let env1 = foldl checkDecls initE p
-              env2 = foldl checkReferencess env1 p
-              in
-            env2 { subObjs = reverse $ subObjs env2 }
-          where initE = SubEnv { subObjs = [], subDefs = M.empty,  subSymbols = M.empty, subApps = [], subArgs = M.empty }
-
-applyDef (n, m) d = case M.lookup n d of
+-- AST as the input and the environment recived from the DSLL type checker, checks the validity of the program acoording to
+-- the typechecking rules, and outputs a collection of information.
+check :: SubProg -> VarEnv -> VarEnv
+check p varEnv = let env = foldl checkSubStmt varEnv p
+                 in if (null (errors env))
+                    then env
+                    else error ("Substance type checking failed with the following problems: \n" ++ (errors env))
+
+checkSubStmt :: VarEnv -> SubStmt -> VarEnv
```

Comment:
add a comment somewhere describing how the various env functions handle multiple declarations with same name, and multiple bindings with same name